### DISCUSSION TITLE
Internal of openVAF

### DISCUSSION THREAD

**Discussion by Lang0722** [pascalkuthe/OpenVAF#130] [Q&A] [answered]
The "Compiler Internal" page on the OpenVAF website is empty. I'm interested in learning more about OpenVAF and improve the analog event control function. Could anyone share more info? I've seen some details of "how openVAF works" in Issues and dissucssion page when author is answering some questions, but I'm looking for more in-depth information.

**Comment by metroid120** [comment, marked as answer]
Hi, you correctly observed that there is no good documentation. This was planed originally, but the main developer of the project has stopped contributing to the project. So the situation is not gonna change anytime soon.

It would be greatly appreciated if you spend the time to work into the project and help fix some issues.

Mabye the original paper holds some hints for you, but it is quite **outdated**:
https://ieeexplore.ieee.org/abstract/document/9193978

To understand this project, you need to have excellent Rust and compiler construction know-how.

**Reply by Lang0722** [reply]
Thank you for the information. It is unfortunate to see such a great project in limbo. After watching the [workshop meeting and presentations](https://www.youtube.com/watch?v=4Cf00ZeueNc) by Ken Kundert and Pascale Kuthe, I believe this project deserves more of our attention.

**Reply by Lang0722** [reply]
> Hi, you correctly observed that there is no good documentation. This was planed originally, but the main developer of the project has stopped contributing to the project. So the situation is not gonna change anytime soon.
> 
> It would be greatly appreciated if you spend the time to work into the project and help fix some issues.
> 
> Mabye the original paper holds some hints for you, but it is quite **outdated**: https://ieeexplore.ieee.org/abstract/document/9193978
> 
> To understand this project, you need to have excellent Rust and compiler construction know-how.

I really tried to understand the project. The frontend part of HIR is fairly understandable, as there is quite a bit of material available from rust-analyzer. However, when it comes to MIR or OSDI, it's really difficult to grasp what the code is doing. There are hardly any comments in the code and no documentation, so I just ended up getting lost while trying to navigate through it.

**Reply by Kreijstal** [reply]
maybe use this fork https://github.com/arpadbuermen/OpenVAF

**Reply by Lang0722** [reply]
> maybe use this fork https://github.com/arpadbuermen/OpenVAF

Thank you for your suggestion. I've been closely following the recent updates across the OpenVAF forks, but unfortunately, progress on backend documentation has been slow. My limited experience and knowledge with Rust and compiler design has made it challenging to dive deeper into OpenVAF’s backend, which appears to combination of Cranelift and LLVM. Lately, I’ve shifted my focus toward C code generation, as it’s more feasible for me to implement additional VA features such as event control, parameter sets, and module reuse. My plan is to leverage OpenVAF’s frontend to generate C-compatible data structures and use cdylib to integrate with my C backend code.
